### ISSUE TITLE
このプロジェクトは貢献を受け取りますか

### ISSUE THREAD

**Issue by AlphaModder** [ekicyou/shiori3-rs#1]
初めまして。私もRustでSHIORIのライブラリを作っていて、このプロジェクトを気づきました。私のものより完成していますので、手伝ったらいいと思っています。今、マクロでSHIORIのDLLを作る方式と始まりのレスポンスのAPIがあります。

変な日本語ですみません。日本語を習っているアメリカ人です。もし、その方が良ければ、英語で話すことも出来ます。

**Comment by ekicyou** [comment]
興味を持ってくれてありがとう！日本語大丈夫ですよ。

差し支えなければ、実験プロジェクトの
https://github.com/ekicyou/emo-rs を見てもらうのも良いかもしれません。
これはスクリプト言語にluaを利用した、Rust SHIORIの実装です。
（今現在、動きません）

元々こちらを作りつつSHIORI APIを切り出そうとしていましたが、
rustやfailureクレートの理解が十分でなく、ライブラリとしての実装に迷いがある状態です。
動くものを作ることが優先目標ですので、現在はemoの制作を優先し、
SHIORIとしてのAPIを整理したいと考えています。

shiori3-rsの方は、emoの成果を使ってAPIの切り出し方を大きく変えたいと考えているため、
今は継続的な改良を受け入れるのは難しいです。emo-rsについては貢献の受け入れは歓迎です。

いずれにせよ私のコードも破壊へ的変更が多発する実験段階ですので、
お互いにコードを見せ合うくらいの方が良いかもしれません。
@AlphaModder さんのコードも、ぜひ参考にさせてください。

**Comment by AlphaModder** [comment]
そうですか。emo-rsを見せてくれてありがとう。私もluaを使うSHIORIの実装を作る定がありました。だから、喜んでemo-rsをお手伝いします。

私のコードは、最近作り始めたプロトタイプだけですので、あまり有用じゃないかもしれません。だけど、ekicyouさんが自分で決めるために、私はもうすぐGitHubにアップロードします。

**Comment by AlphaModder** [comment]
はい、[https://github.com/AlphaModder/rust-shiori](https://github.com/AlphaModder/rust-shiori)にアップロードしました。`shiori!`のマクロは最も興味がある部分でしょう。

**Comment by ekicyou** [comment]
プライベートが忙しくなるため、しばらく反応が鈍くなります。
shiori! マクロ周辺読みました。rustはインターフェースが作りにくいと感じていましたが、マクロで分離するのはいい考えだと思いました。インターフェース整理のタイミングで取り込みを検討します。

以下、気になったところのみリストします。
1. load apiのpathですが、ANSI文字列です。UTF-8決め打ちですと日本語環境で問題が出ます。
2. load apiのpathはHGLOBAL渡しですので、load側でGlobalFreeが必要です。

**Comment by AlphaModder** [comment]
反応の速度は心配しないでください。私も忙しかったのです。`shiori!`は役に立ってよかったです。気軽にお使いください。

`GlobalFree`は確かに忘れてしまいました。指摘してくれてありがとう。エンコーディングの問題には、ekicyouさんの`GStr`の実装を読みました。私が使ってもいいですか。

ところで、`GStr::as_bytes`と`GStr::new`の実装を読むとき、`mem::transmute`で`HGLOBAL`から`*mut u8`に変換することに気付きました。間違っていないが、標準的な方法は`h as *mut u8`です。
